File: strings/49. Group Anagrams.py
use std::collections::HashMap;

impl Solution {
    pub fn group_anagrams(strs: Vec<String>) -> Vec<Vec<String>> {
        let mut map = HashMap::<Vec<u8>, Vec<String>>::new();
        for curr_str in strs {
            let mut freq = vec![0u8; 26];

            for ch in curr_str.chars() {
                freq[ch as usize - 'a' as usize] += 1
            }

            map.entry(freq)
            .or_insert(Vec::new())
            .push(curr_str);
        }
        map.into_values().collect()
    }
}

File: strings/49. Group Anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string,vector<string>> map;
        for(const auto& str : strs) {
            vector<int> freq(26, 0);
            for(char ch : str) {
                freq[ch - 'a']++;
            }
            string key = to_string(freq[0]);
            for(int i = 1; i < 26; i++) {
                key += ',' + to_string(freq[i]);
            }
            map[key].push_back(str);
        }
        vector<vector<string>> res;
        for(const auto& freq : map) {
            res.push_back(freq.second);
        }
        return res;
    }
};
